tpu-client-next/src/node_address_service/recent_leader_slots.rs

```diff
@@ -0,0 +1,177 @@
+//! This module provides [`RecentLeaderSlots`] to track recent leader slots.
+use {
+    crate::node_address_service::slot_receiver::EstimatedSlot, log::*, solana_clock::Slot,
+    std::collections::VecDeque,
+};
+
+// 48 chosen because it's unlikely that 12 leaders in a row will miss their slots
+const MAX_SLOT_SKIP_DISTANCE: u64 = 48;
+
+const RECENT_LEADER_SLOTS_CAPACITY: usize = 48;
+
+#[derive(Debug, Clone)]
+struct SlotEvent {
+    slot: Slot,
+    is_start: bool,
+}
+
+#[derive(Debug)]
+pub struct RecentLeaderSlots(VecDeque<SlotEvent>);
+
+impl RecentLeaderSlots {
+    pub fn new() -> Self {
+        Self(VecDeque::with_capacity(RECENT_LEADER_SLOTS_CAPACITY))
+    }
+}
+
+impl Default for RecentLeaderSlots {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl RecentLeaderSlots {
+    pub fn record_slot_start(&mut self, current_slot: Slot) {
+        self.0.push_back(SlotEvent {
+            slot: current_slot,
+            is_start: true,
+        });
+        while self.0.len() > RECENT_LEADER_SLOTS_CAPACITY {
+            self.0.pop_front();
+        }
+    }
+
+    pub fn record_slot_end(&mut self, current_slot: Slot) {
+        self.0.push_back(SlotEvent {
+            slot: current_slot,
+            is_start: false,
+        });
+        while self.0.len() > RECENT_LEADER_SLOTS_CAPACITY {
+            self.0.pop_front();
+        }
+    }
+
+    // Estimate the current slot from recent slot notifications.
+    #[allow(clippy::arithmetic_side_effects)]
+    pub fn estimate_current_slot(&self) -> EstimatedSlot {
+        let mut recent_slots: Vec<SlotEvent> = self.0.iter().cloned().collect();
+        assert!(!recent_slots.is_empty());
+        recent_slots.sort_by(|a, b| {
+            a.slot
+                .cmp(&b.slot)
+                .then_with(|| b.is_start.cmp(&a.is_start)) // true before false
+        });
```

Comment:
Performance is not as important for this code, but still.
Would it be perhaps better to just keep slot events ordered in [a priority queue](https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html)?